src/function/matrix/eigs.js

```diff
@@ -41,63 +54,69 @@ export const createEigs = /* #__PURE__ */ factory(name, dependencies, ({ config,
    * @param {Array | Matrix} x  Matrix to be diagonalized
    *
    * @param {number | BigNumber} [prec] Precision, default value: 1e-15
-   * @return {{values: Array|Matrix, vectors: Array|Matrix}} Object containing an array of eigenvalues and a matrix with eigenvectors as columns.
+   * @return {{values: Array|Matrix, eigenvectors: Array<EVobj>}} Object containing an array of eigenvalues and an array of {value: number|BigNumber, vector: Array|Matrix} objects.
    *
    */
   return typed('eigs', {
 
-    Array: function (x) {
-      const mat = matrix(x)
-      return computeValuesAndVectors(mat)
-    },
-
+    // The conversion to matrix in the first two implementations,
+    // just to convert back to an array right away in
+    // computeValuesAndVectors, is unfortunate, and should perhaps be
+    // streamlined. It is done because the Matrix object carries some
+    // type information about its entries, and so constructing the matrix
+    // is a roundabout way of doing type detection.
+    Array: function (x) { return computeValuesAndVectors(matrix(x)) },
     'Array, number|BigNumber': function (x, prec) {
-      const mat = matrix(x)
-      return computeValuesAndVectors(mat, prec)
+      return computeValuesAndVectors(matrix(x), prec)
     },
-
     Matrix: function (mat) {
-      const { values, vectors } = computeValuesAndVectors(mat)
-      return {
-        values: matrix(values),
-        vectors: matrix(vectors)
-      }
+      return computeValuesAndVectors(mat, undefined, true)
     },
-
     'Matrix, number|BigNumber': function (mat, prec) {
-      const { values, vectors } = computeValuesAndVectors(mat, prec)
-      return {
-        values: matrix(values),
-        vectors: matrix(vectors)
-      }
+      return computeValuesAndVectors(mat, prec, true)
     }
   })
 
-  function computeValuesAndVectors (mat, prec) {
+  function computeValuesAndVectors (mat, prec, matricize = false) {
     if (prec === undefined) {
       prec = config.epsilon
     }
 
-    const size = mat.size()
+    let result
+    const arr = mat.toArray()
+    const asize = mat.size()
 
-    if (size.length !== 2 || size[0] !== size[1]) {
-      throw new RangeError('Matrix must be square (size: ' + format(size) + ')')
+    if (asize.length !== 2 || asize[0] !== asize[1]) {
+      throw new RangeError(`Matrix must be square (size: ${format(asize)})`)
     }
 
-    const arr = mat.toArray()
-    const N = size[0]
+    const N = asize[0]
 
     if (isReal(arr, N, prec)) {
       coerceReal(arr, N)
 
       if (isSymmetric(arr, N, prec)) {
         const type = coerceTypes(mat, arr, N)
-        return doRealSymetric(arr, N, prec, type)
+        result = doRealSymmetric(arr, N, prec, type)
       }
     }
-
-    const type = coerceTypes(mat, arr, N)
-    return doComplexEigs(arr, N, prec, type)
+    if (!result) {
+      const type = coerceTypes(mat, arr, N)
+      result = doComplexEigs(arr, N, prec, type)
+    }
+    if (matricize) {
```

Comment:
I'm personally not a big fan of having a mutable variable like `result` that is changed inside nested if's a few times, it makes it hard to follow the logic of what states `result` may have. 

Is it possible to extract this logic in two helper functions `matricize` and `attachVectorsError`, or a single one `prepareOutput` or so, and use that like `return prepareOutput(matricize(computeValuesAndVectors(mat, undefined)))`? Then you could keep this plumbing out of `computeValuesAndVectors` and basically keep it as it was with a direct `return` of the result.

Reply:
OK, I _think_ I did the refactor you wanted. If this still isn't organized the way you like, please suggest another refactor and I am happy to make it so.

Reply:
I like that 👍